api/src/main/java/graphql/nadel/normalized/ValueToVariableValueCompiler.java

```diff
@@ -0,0 +1,107 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+public class ValueToVariableValueCompiler {
+
+    static VariableValueWithDefinition normalizedInputValueToVariable(NormalizedInputValue normalizedInputValue) {
+        Object variableValue = null;
+        Object normalizedInputValueValue = normalizedInputValue.getValue();
+        if (normalizedInputValueValue instanceof ObjectValue) {
+            variableValue = toVariableValue((ObjectValue) normalizedInputValueValue);
+        }
+        if (normalizedInputValueValue instanceof ArrayValue) {
+            variableValue = toVariableValues(((ArrayValue) normalizedInputValueValue).getValues());
+        }
+        if (normalizedInputValueValue instanceof List) {
+            variableValue = toVariableValues((List) normalizedInputValueValue);
+        }
+        String varName = getVarName();
+        return new VariableValueWithDefinition(
+                variableValue,
+                VariableDefinition.newVariableDefinition()
+                        .name(varName)
+                        .type(TypeName.newTypeName(normalizedInputValue.getTypeName()).build())
+                        .build(),
+                VariableReference.newVariableReference().name(varName).build());
+    }
+
+    @NotNull
+    private static List<Object> toVariableValues(List<Value> values) {
+        ArrayList<Object> objects = new ArrayList<>();
+        for (Value value : values) {
+            objects.add(toVariableValue(value));
+        }
+        return objects;
+    }
+
+    private static Map<String, Object> toVariableValue(ObjectValue objectValue) {
+        HashMap<String, Object> map = new HashMap<>();
+        List<ObjectField> objectFields = objectValue.getObjectFields();
+        for (ObjectField objectField : objectFields) {
+            String objectFieldName = objectField.getName();
+            Value<?> objectFieldValue = objectField.getValue();
+            if (objectFieldValue instanceof ArrayValue) {
+                List<Object> objects = toVariableValues(((ArrayValue) objectFieldValue).getValues());
+                map.put(objectFieldName, objects);
+                continue;
+            }
+            map.put(objectFieldName, toVariableValue(objectFieldValue));
+        }
+        return map;
+    }
+
+    private static Object toVariableValue(Value value) {
+        if (value instanceof ObjectValue) {
+            return toVariableValue((ObjectValue) value);
+        }
+        if (value instanceof StringValue) {
+            return ((StringValue) value).getValue();
+        }
+        if (value instanceof FloatValue) {
+            return ((FloatValue) value).getValue();
+        }
+        if (value instanceof IntValue) {
+            return ((IntValue) value).getValue();
+        }
+        if (value instanceof BooleanValue) {
+            return ((BooleanValue) value).isValue();
+        }
+        return null; //todo
```

Comment:
need to double check all possible types here

Reply:
Here's a list of `Value` implementations according to IntelliJ

ArrayValue
BooleanValue
EnumValue
FloatValue
IntValue
NullValue
ObjectValue
ScalarValue
StringValue
VariableReference

Once you're done with that, we should throw an exception at the bottom instead of returning null. We wouldn't want to silently send a malformed query.

Reply:
+1 for throwing an exception at the end.
Some scalars are not relevant as we rely here on a specific JSON implementation: there will never be a VariableReference for example.

Reply:
so I believe EnumValue, ScalarValue, VariableReference are the only ones that are not relevant

Reply:
these types are never returned from https://github.com/graphql-java/graphql-java-extended-scalars/blob/master/src/main/java/graphql/scalars/object/ObjectScalar.java#L107